src/Symfony/Component/Routing/RouteCompiler.php

```diff
@@ -174,6 +174,192 @@ private function compilePattern(Route $route, $pattern, $isHostname)
     }
 
     /**
+     * Performs lexical and syntactic analysis of the pattern and returns the parse tree
+     * consisting of tokens.
+     *
+     * The array is a tree when optional parts enclosed with parentheses are used
+     * that can potentially be nested at any deepth.
+     *
+     * @param string  $pattern     The route pattern or subpattern for an optional part
+     * @param Boolean $isHostname  Whether it is the pattern for the hostname or path
+     * @param string  $fullPattern The full pattern used for better exception messages
+     * @param integer $parentPos   The parsing position of the parent call
+     *
+     * @return array The parse tree of tokens
+     *
+     * @throws \LogicException If the pattern is invalid
+     */
+    public static function parsePattern($pattern, $isHostname = false, $fullPattern = '', $parentPos = 0)
+    {
+        $tokens = array();
+        $matches = array();
+        $pos = 0;
+
+        if ('' === $fullPattern) {
+            $fullPattern = $pattern;
+        }
+
+        // '#\{.*?\}|\((?:[^()]++|(?R))*\)#'
+        // '#(?<!\\\\)\{.*?(?<!\\\\)\}|(?<!\\\\)\((?:[^()]++|((?<=\\\\)[()])++|(?R))*(?<!\\\\)\)#'
+        preg_match_all('#(?<!\\\\)\{.*?(?<!\\\\)\}|(?<!\\\\)\((?:[^()]++|((?<=\\\\)[()])++|(?R))*(?<!\\\\)\)#', $pattern, $matches, PREG_OFFSET_CAPTURE | PREG_SET_ORDER);
```

Comment:
@Tobion if you can use ``[^()]++``, it is much better as it prevents useless backtracking. This makes the parsing both faster and less likely to break on long string in case it cannot match.